files/worker/wfexec/event-timetable.pl

```diff
@@ -28,325 +31,274 @@
-  my $series = shift;
-  my $result = "";
+    my ($visibility, $course, $caption_provider) = ("vula", "", "");
 
-  $mech->get("$server/series/$series/acl.json");
-  my $response = $mech->response();
-  if (!$response->is_success) {
-    $timetable_completed = 0;
-    #$timetable_result = "Unable to get series ACL for series $series: " . $response->status_line;
-    die "Unable to get series ACL for series $series: " . $response->status_line;
-  } else {
-    $result = $response->decoded_content;
-  }
+    my $series_acl = wrapGetMetadata($mech, "$server/series/$series/acl.json", 0) =~ /ROLE_ANONYMOUS/;
+    if ($series_acl) {
+        $visibility = "public";
+    }
+
+    my $meta = wrapGetMetadata($mech, "$server/api/series/$series/metadata", 1);
+    if (defined($meta)) {
+        $course = getSeriesField($meta, "course");
+        $caption_provider = getSeriesField($meta, "caption-type");
+    }
 
-  return $result;
+    return ($visibility, $course, $caption_provider)
 }
 
-# Series Metadata
+# Event Metadata
+sub getEventDetails($$$) {
 
-sub getSeriesMetadata($$) {
+    my $mech = shift;
+    my $server = shift;
+    my $mp = shift;
 
-  my $mech = shift;
-  my $server = shift;
-  my $series = shift;
-  my $result = "";
+    ## Get the mediapackage info from api / admin-ng
+    my ($series, $event_date, $duration, $title, $start_time, $end_time) = ("","","","","","");
+    my $event_m = wrapGetMetadata($mech, "$server/api/events/$mp/metadata?type=dublincore%2Fepisode", 1);
 
-  $mech->get("$server/api/series/$series/metadata");
-  my $response = $mech->response();
-  if (!$response->is_success) {
-    $timetable_completed = 0;
-    #$timetable_result = "Unable to get series metadata for series $series: " . $response->status_line;
-    die "Unable to get series metadata for series $series: " . $response->status_line;
-  } else {
-    $result = $response->decoded_content;
-  }
+    $series     = getEventField($event_m, "isPartOf", 0);
+    $event_date = $w3c->parse_datetime(getEventField($event_m, "startDate", 0) ."T". getEventField($event_m, "startTime", 0) ."Z");
+    $duration   = getEventField($event_m, "duration", 0);
+    $title      = getEventField($event_m, "title", 0);
+    $location   = getEventField($event_m, "location", 0);
 
-  return $result;
-}
+    my $local_time = $event_date->clone()->set_time_zone('GMT')->set_time_zone('Africa/Johannesburg');
+    my $local_end_time = $local_time + $dfd_t->parse_duration($duration);
 
-# Event Metadata
-# /api/events/MPID?withacl=true&withmetadata=true to get location, start, duration, series (isPartOf), ACL, e.g.:
+    $start_time = $local_time->strftime("%H:%M");
+    $event_date = $local_time->strftime("%Y-%m-%d");
+    $end_time   = $local_end_time->strftime("%H:%M");
 
-sub getEventMetadata($$) {
+    return ($series, $event_date, $duration, $title, $start_time, $end_time);
+}
 
-  my $mech = shift;
-  my $server = shift;
-  my $mp = shift;
+sub wrapGetMetadata($$$) {
```

Comment:
This whole code structure is dodgy, especially setting a global var inside a method like this. Much cleaner just to implement the retry as a for or while inside getMetadata instead of having a wrapper function to call a function using recursion for retries

Reply:
Here is an example of a cleaner implementation of the same retry concept using recursion:

https://github.com/opencast/opencast/commit/fbf1c192bc4f68c6199e338bc3b0b3770c8eda59#diff-60b0134db7179ba1569021d0853c1004

Pass the retry count as a parameter, avoiding a global var and a wrapper function.